File: accounts-go/account/account.go
package account

import (
	"errors"
	"fmt"
)

type AggregateId UUID
type OwnerId UUID

type account struct {
	id      *AggregateId
	ownerId *OwnerId
	balance int64
	open    bool
}

type AccountSnapshot struct {
	id      AggregateId
	ownerId OwnerId
	balance int64
	open    bool
}

func NewAccount() *account {
	return &account{}
}

func (a account) Id() AggregateId {
	return *a.id
}

func (a *account) Snapshot() (*AccountSnapshot, error) {
	if !a.open {
		return nil, errors.New("account not open")
	}

	return &AccountSnapshot{*a.id, *a.ownerId, a.balance, a.open}, nil
}

func (a *account) Open(accountId AggregateId, ownerId OwnerId) (Event, error) {
	if a.id != nil || a.ownerId != nil {
		return nil, errors.New("account already open")
	}

	event := AccountOpenedEvent{accountId, ownerId}
	a.applyAccountOpened(event)
	return event, nil
}

func (a *account) Deposit(amount int64) (Event, error) {
	if amount < 0 {
		return nil, errors.New("Can not deposit negative amount")
	}
	if !a.open {
		return nil, errors.New("account not open")
	}
	if amount == 0 {
		return nil, nil
	}

	event := MoneyDepositedEvent{amount, a.balance + amount}
	a.applyMoneyDeposited(event)
	return event, nil
}

func (a *account) applyAccountOpened(event AccountOpenedEvent) {
	a.id = &event.accountId
	a.ownerId = &event.ownerId
	a.balance = 0
	a.open = true
}

func (a *account) applyMoneyDeposited(event MoneyDepositedEvent) {
	a.balance = event.balance
}

type Event interface {
	apply(account *account)
	//Serialize() []byte
}

type AccountOpenedEvent struct {
	accountId AggregateId
	ownerId   OwnerId
}

func (e AccountOpenedEvent) String() string {
	return fmt.Sprintf("AccountOpenedEvent{ownerId: %s}", e.ownerId)
}

func (e AccountOpenedEvent) apply(account *account) {
	account.applyAccountOpened(e)
}

type MoneyDepositedEvent struct {
	amountDeposited int64
	balance         int64
}

func (e MoneyDepositedEvent) apply(account *account) {
	account.applyMoneyDeposited(e)
}

/*func (e AccountOpenedEvent) Serialize() []byte {
	return e.ownerId
}*/

File: src/main/java/lt/rieske/accounts/domain/Account.java
package lt.rieske.accounts.domain;

import lt.rieske.accounts.eventsourcing.EventStream;

import java.util.UUID;


public class Account implements AccountEventsVisitor {
    private final EventStream<Account, AccountEventsVisitor> eventStream;

    private final UUID accountId;

    private UUID ownerId;
    private long balance;
    private boolean open;

    public Account(EventStream<Account, AccountEventsVisitor> eventStream, UUID accountId) {
        this.eventStream = eventStream;
        this.accountId = accountId;
    }

    public AccountSnapshot<AccountEventsVisitor> snapshot() {
        return new AccountSnapshot<>(accountId, ownerId, balance, open);
    }

    public void open(UUID ownerId) {
        if (this.ownerId != null) {
            throw new IllegalStateException("Account already has an owner");
        }
        eventStream.append(new AccountOpenedEvent<>(ownerId), this, accountId);
    }

    public void deposit(long amount) {
        if (amount == 0) {
            return;
        }
        requireOpenAccount();
        if (amount < 0) {
            throw new IllegalArgumentException("Can not deposit negative amount: " + amount);
        }
        eventStream.append(new MoneyDepositedEvent<>(amount, balance + amount), this, accountId);
    }

    public void withdraw(long amount) {
        if (amount == 0) {
            return;
        }
        requireOpenAccount();
        if (amount < 0) {
            throw new IllegalArgumentException("Can not withdraw negative amount: " + amount);
        }
        if (balance < amount) {
            throw new IllegalArgumentException("Insufficient balance");
        }
        eventStream.append(new MoneyWithdrawnEvent<>(amount, balance - amount), this, accountId);
    }

    public void close() {
        if (balance != 0) {
            throw new IllegalStateException("Balance outstanding");
        }
        eventStream.append(new AccountClosedEvent<>(), this, accountId);
    }

    public UUID id() {
        return accountId;
    }

    public UUID ownerId() {
        return ownerId;
    }

    public long balance() {
        return balance;
    }

    boolean isOpen() {
        return open;
    }

    @Override
    public void visit(AccountSnapshot snapshot) {
        this.ownerId = snapshot.ownerId();
        this.balance = snapshot.balance();
        this.open = snapshot.open();
    }

    @Override
    public void visit(AccountOpenedEvent event) {
        this.ownerId = event.ownerId();
        this.balance = 0;
        this.open = true;
    }

    @Override
    public void visit(MoneyDepositedEvent event) {
        this.balance = event.balance();
    }

    @Override
    public void visit(MoneyWithdrawnEvent event) {
        this.balance = event.balance();
    }

    @Override
    public void visit(AccountClosedEvent event) {
        this.open = false;
    }

    private void requireOpenAccount() {
        if (!open) {
            throw new IllegalStateException("Account not open");
        }
    }

}
